File: IntroToJava/Book.java
// Book.java
// Base class representing a novel or general book entry in the catalog.
// All other entry types inherit from this class.

public class Book {

    // Fields common to all catalog entries
    private String type;
    private String title;
    private String author;
    private String authorGender;
    private String publisher;
    private double price;
    private boolean readStatus;

    // Constructor: initializes all fields for a standard book entry
    public Book(String title, String author, String authorGender,
                String publisher, double price, boolean readStatus) {
        this.type = "novel";
        this.title = title;
        this.author = author;
        this.authorGender = authorGender;
        this.publisher = publisher;
        this.price = price;
        this.readStatus = readStatus;
    }

    // Protected constructor used by subclasses to set a different type
    protected Book(String type, String title, String author, String authorGender,
                   String publisher, double price, boolean readStatus) {
        this.type = type;
        this.title = title;
        this.author = author;
        this.authorGender = authorGender;
        this.publisher = publisher;
        this.price = price;
        this.readStatus = readStatus;
    }

    // Prints all fields of this entry in a readable format
    public void display() {
        String status = readStatus ? "Yes" : "No";
        System.out.println("\n  Type       : Novel");
        System.out.println("  Title      : " + title);
        System.out.println("  Author     : " + author + " (" + authorGender + ")");
        System.out.println("  Publisher  : " + publisher);
        System.out.printf ("  Price      : %.2f EUR%n", price);
        System.out.println("  Read       : " + status);
        System.out.println("  " + "-".repeat(40));
    }

    // Sets the read status to true
    public void markAsRead() {
        this.readStatus = true;
    }

    // Updates a field by name. Returns true if the update succeeded.
    public boolean update(String field, String newValue) {
        switch (field) {
            case "title":
                this.title = newValue;
                break;
            case "author":
                this.author = newValue;
                break;
            case "authorGender":
                this.authorGender = newValue;
                break;
            case "publisher":
                this.publisher = newValue;
                break;
            case "price":
                // Price must be a valid number; reject non-numeric input
                try {
                    this.price = Double.parseDouble(newValue);
                } catch (NumberFormatException e) {
                    System.out.println("  Invalid price. Please enter a number.");
                    return false;
                }
                break;
            case "readStatus":
                // Accept common truthy/falsy strings
                this.readStatus = newValue.equalsIgnoreCase("yes")
                               || newValue.equalsIgnoreCase("true")
                               || newValue.equals("1");
                break;
            default:
                System.out.println("  Field '" + field + "' not found.");
                return false;
        }
        System.out.println("  Field '" + field + "' updated successfully.");
        return true;
    }

    // Converts this entry to a pipe-separated string for file storage
    public String toFileLine() {
        return type + "|" + title + "|" + author + "|" + authorGender + "|"
             + publisher + "|" + price + "|" + readStatus;
    }

    // Getters
    public String getType()         { return type; }
    public String getTitle()        { return title; }
    public String getAuthor()       { return author; }
    public String getAuthorGender() { return authorGender; }
    public String getPublisher()    { return publisher; }
    public double getPrice()        { return price; }
    public boolean getReadStatus()  { return readStatus; }

    // Allows subclasses to override the type field
    protected void setType(String type) { this.type = type; }
}

File: IntroToPython/Book.py
# Book.py 

# This file defines the Book base class and the Comic subclass.
# Book represents a novel or any general book.
# Comic inherits from Book and adds fields specific to illustrated works,
# such as illustrator, illustrator gender, and visual style.


class Book:
    """Base class representing a novel or general book."""

    def __init__(self, title, author, author_gender, publisher, price, read_status):
        self.type = "novel"
        self.title = title
        self.author = author
        self.author_gender = author_gender
        self.publisher = publisher
        self.price = price
        self.read_status = read_status

    def display(self):
        """Print all book attributes in a readable format."""
        status = "Yes" if self.read_status else "No"
        print(f"\n  Type       : Novel")
        print(f"  Title      : {self.title}")
        print(f"  Author     : {self.author} ({self.author_gender})")
        print(f"  Publisher  : {self.publisher}")
        print(f"  Price      : {self.price} EUR")
        print(f"  Read       : {status}")
        print("  " + "-" * 40)

    def mark_as_read(self):
        """Set the read status to True."""
        self.read_status = True

    def update(self, field, new_value):
        """Update a field if it exists on this object."""
        valid_fields = ["title", "author", "author_gender", "publisher", "price", "read_status"]
        if field in valid_fields:
            if field == "price":
                try:
                    new_value = float(new_value)
                except ValueError:
                    print("  Invalid price. Please enter a number.")
                    return
            if field == "read_status":
                new_value = new_value.lower() in ["yes", "true", "1"]
            setattr(self, field, new_value)
            print(f"  Field '{field}' updated successfully.")
        else:
            print(f"  Field '{field}' not found.")

    def to_dict(self):
        """Convert the object to a dictionary for JSON serialization."""
        return {
            "type": self.type,
            "title": self.title,
            "author": self.author,
            "author_gender": self.author_gender,
            "publisher": self.publisher,
            "price": self.price,
            "read_status": self.read_status
        }


class Comic(Book):
    """Subclass representing a comic, graphic novel, or manga."""

    def __init__(self, title, author, author_gender, publisher, price, read_status,
                 illustrator, illustrator_gender, style):
        super().__init__(title, author, author_gender, publisher, price, read_status)
        self.type = "comic"
        self.illustrator = illustrator
        self.illustrator_gender = illustrator_gender
        self.style = style

    def display(self):
        """Print all comic attributes, including illustrator and style."""
        status = "Yes" if self.read_status else "No"
        print(f"\n  Type       : Comic")
        print(f"  Title      : {self.title}")
        print(f"  Author     : {self.author} ({self.author_gender})")
        print(f"  Illustrator: {self.illustrator} ({self.illustrator_gender})")
        print(f"  Style      : {self.style}")
        print(f"  Publisher  : {self.publisher}")
        print(f"  Price      : {self.price} EUR")
        print(f"  Read       : {status}")
        print("  " + "-" * 40)

    def update_illustrator(self, field, new_value):
        """Update a Comic-specific field."""
        valid_fields = ["illustrator", "illustrator_gender", "style"]
        if field in valid_fields:
            setattr(self, field, new_value)
            print(f"  Field '{field}' updated successfully.")
        else:
            print(f"  Field '{field}' is not a valid comic-specific field.")

    def to_dict(self):
        """Convert the Comic object to a dictionary for JSON serialization."""
        data = super().to_dict()
        data["illustrator"] = self.illustrator
        data["illustrator_gender"] = self.illustrator_gender
        data["style"] = self.style
        return data
